### DISCUSSION TITLE
Cant connect to MySQL databse after upgrade to 1.18.0

### DISCUSSION THREAD

**Discussion by EC58** [dani-garcia/vaultwarden#1348] [Q&A]
### Subject of the issue
After updating my docker version from 1.17 to 1.18 I can no longer connect to the MySQL database:

Getting the following error:
[bitwarden_rs][ERROR] Error creating database pool: DieselConError.

[CAUSE] BadConnection(

    "Plugin caching_sha2_password could not be loaded: /usr/lib/x86_64-linux-gnu/mariadb19/plugin/caching_sha2_password.so: cannot open shared object file: No such file or directory",

)

My mysql user is using mysql_native_password so I dont see why this error pops up. Should I be using MariaDB instead?

**Comment by musover** [comment]
I am also having this issue, running in a kubernetes cluster using CRI-O as a container runtime.

**Comment by BlackDex** [comment]
I suggest to use a mariadb container for the database. Or create a database user which does not use the the latest mysql password features. That error message will only appear if it needs to use that password feature. But that isn't build into. It is something we need to see if that is possible at all using the current library versions.

**Comment by sbmw** [comment]
Thanks @BlackDex but (as with @EC58) my database user is already set to not use the sha2 password features.  The same db user was working fine with 1.17.0. 

```
mysql> select User, Plugin from user;

+------------------+-----------------------+
| User             | Plugin                |
+------------------+-----------------------+
| bitwarden        | mysql_native_password |

```

Perhaps one of the updated dependencies in 1.18.0? Or could it be trying to run a migration with a user (eg root) other than I have set database_url in config?

 "database_url": "mysql://bitwarden:PASSWORD@mysql/bitwarden",

**Comment by BlackDex** [comment]
Which image are you using? amd64? Alpine or default?

**Comment by sbmw** [comment]
Default I believe:

bitwardenrs/server-mysql:1.18.0

**Comment by BlackDex** [comment]
I would suggest to start using the `bitwardenrs/server:latest` version with the next release instead of the `-mysql`, both are the same and we will probably remove the -mysql and -postgresql versions in the future.

Could you try to install the libmariadb3 package within the docker?
It will only work untill the next restart, but at least you could test if it works.

You could add this via the start.sh feature by creating a script called bitwarden_rs.sh and volume-mount that to /etc/bitwarden_rs.sh which gets started before bitwarden itself.

Check if that works.

**Comment by sbmw** [comment]
Hey, so I switched to bitwardenrs/server:latest and volume mounted the startup file with contents:

```
apt-get update
apt-get install libmariadb3
```

The container log now shows:

```
Get:1 http://security.debian.org/debian-security buster/updates InRelease [65.4 kB]
Get:2 http://security.debian.org/debian-security buster/updates/main amd64 Packages [260 kB]
Get:3 http://deb.debian.org/debian buster InRelease [121 kB]
Get:4 http://deb.debian.org/debian buster-updates InRelease [51.9 kB]
Get:5 http://deb.debian.org/debian buster/main amd64 Packages [7907 kB]
Get:6 http://deb.debian.org/debian buster-updates/main amd64 Packages [7860 B]
Fetched 8414 kB in 2s (4910 kB/s)
Reading package lists...
Reading package lists...
Building dependency tree...
Reading state information...
libmariadb3 is already the newest version (1:10.3.27-0+deb10u1).
libmariadb3 set to manually installed.
0 upgraded, 0 newly installed, 0 to remove and 3 not upgraded.
/--------------------------------------------------------------------\
|                       Starting Bitwarden_RS                        |
|                           Version 1.18.0                           |
|--------------------------------------------------------------------|
| This is an *unofficial* Bitwarden implementation, DO NOT use the   |
| official channels to report bugs/features, regardless of client.   |
| Send usage/configuration questions or feature requests to:         |
|   https://bitwardenrs.discourse.group/                             |
| Report suspected bugs/issues in the software itself at:            |
|   https://github.com/dani-garcia/bitwarden_rs/issues/new           |
\--------------------------------------------------------------------/

[INFO] No .env file found.

[2021-01-04 12:41:31.018][bitwarden_rs::util][WARN] Can't connect to database, retrying: DieselConError.
[CAUSE] BadConnection(
    "Plugin caching_sha2_password could not be loaded: /usr/lib/x86_64-linux-gnu/mariadb19/plugin/caching_sha2_password.so: cannot open shared object file: No such file or directory",
)

...
```

So, already installed?

**Comment by sbmw** [comment]
Is "[INFO] No .env file found." the issue?

I'm volume mounting a directory containing a config.json to /data, this contains my mysql user details.

Should we be using a .env file?

**Comment by BlackDex** [comment]
No, that is just to notice that your environment isn't using a .env file.
Docker loads all the environment variables into the environment. So nothing to worry about.

The strange thing is.
Debian (Buster) doesn't have that library file at all, it's no where in there repo's.
It only is on Ubuntu, and that is not what we are using to build these container images.
All is using the same Debian version.

**Comment by jjlin** [comment]
This seems unlikely to have anything to do with the bitwarden_rs upgrade per se, as I don't think there have been any related changes between 1.17.0 and 1.18.0. So there's probably been a MySQL server or configuration change on your side, perhaps without your knowledge. If you do a web search for `Plugin caching_sha2_password could not be loaded`, there are various things you can try.

Debian will be including this plugin in the next release (Bullseye), but that probably won't be out until mid-late 2021:

* https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=962597
* https://packages.debian.org/bullseye/amd64/libmariadb3/filelist

**Comment by EC58** [comment]
Could this be the issue: 
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=962597

**Comment by musover** [comment]
While attempting to migrate from MySQL to PostgreSQL, I have found out that despite being configured to use mysql_native_password as an authentication backend, the user bitwarden_rs uses to connect was still using caching_sha2_password regardless, and this is due to a missing setting on /etc/my.cnf on the MySQL server itself.

As a stopgap solution, you can try adding/uncommenting `default-authentication-plugin=mysql_native_password` on the [mysqld] section, which will enforce mysql_native_password.

**Comment by everyx** [comment]
> While attempting to migrate from MySQL to PostgreSQL, I have found out that despite being configured to use mysql_native_password as an authentication backend, the user bitwarden_rs uses to connect was still using caching_sha2_password regardless, and this is due to a missing setting on /etc/my.cnf on the MySQL server itself.
> 
> As a stopgap solution, you can try adding/uncommenting `default-authentication-plugin=mysql_native_password` on the [mysqld] section, which will enforce mysql_native_password.

I have already changed the user's auth type to `mysql_native_password`, but not change the global default auth type to `mysql_native_password`, and  this error is still be reported when upgrade to v1.18.

**Comment by rouben** [comment]
Thank you, @musover, this fixes the issue.
I am running MySQL server version: 8.0.22-0ubuntu0.20.04.3 - (Ubuntu). This is not MariaDB.

@everyx, try the suggestion. This option needs to be added to the appropriate section of your MySQL server, not the Bitwarden docker image.

**Comment by everyx** [comment]
> Thank you, @musover, this fixes the issue.
> I am running MySQL server version: 8.0.22-0ubuntu0.20.04.3 - (Ubuntu). This is not MariaDB.
> 
> @everyx, try the suggestion. This option needs to be added to the appropriate section of your MySQL server, not the Bitwarden docker image.

@rouben I use MySQL 8 instance offered by DigitalOcean Management Database servcie, so I couldn't change the global auth type, but I can changed this for a single user, and I did it, but the error still happens, the v1.17 is worked for this user-specified workaround.

**Comment by rouben** [comment]
> @rouben I use MySQL 8 instance offered by DigitalOcean Management Database servcie, so I couldn't change the global auth type, but I can changed this for a single user, and I did it, but the error still happens, the v1.17 is worked for this user-specified workaround.

@everyx I tried the per-user override as well with no success. The only thing that worked for me is the server-level setting. Unfortunately with a managed MySQL instance you may not have the ability to change that setting. Having said this, I don't know why the previous release of BitWarden-RS MySQL docker build did work with the per-user override. There is speculation that it may have something to do with the build image for the MySQL enabled BitWatden-RS binary within this docker image...

**Comment by demize** [comment]
This should not have been converted from an issue, since this is definitely an issue with the client, not the server (and it's not a documentation issue, either, like implied in #867). As Tim notes in #867, the official MariaDB client is supposed to come with this plugin, so that it's missing is an issue in and of itself, but other software that uses the MariaDB client and is missing this plugin does not have this issue. I'm trying to switch my services over to DigitalOcean and use their managed DB service, and Seafile worked exactly as expected: couldn't use the new authentication method, but changing the user's authentication method in the DO console let it log in fine.

Blaming the server is not the answer here, blaming the documentation is not the answer here, this is absolutely a client issue.

**Comment by BlackDex** [comment]
TL;DR:
Debian does not have this plugin build within there packages, we can't (and will not) try to fix there packages.

If you ensist on using MySQL8 then create a user using `... IDENTIFIED WITH mysql_native_password BY ...`

---

To be blunt, we use the packages provided by the underlying OS, which in this case is Debian.
Debian does **NOT** support/build this new `caching_sha2_password` method! It is not something we will just build in our self by recompiling the default provided packages by Debian, this feature wasn't in there before, but it will be in there eventually.

I think you forget that every distro has there own building flags and which features they add or remove and which versions of those tools they support or not.

As debian doesn't support the latest version of mariadb-client we also can not support this.

I will update the documentation to mention that we do not support MySQL 8.x using it's default password caching and that  you need to use the `mysql_native_password` instead!

Again, this is not something we can and will fix on our side, if Debian releases a new stable version of there Distro we will update that accordingly and we just have to hope it adds this feature.

Also, our builds use the MariaDB libraries since that is what Debian, Ubuntu and Alpine support, there are no packages in there main repo's which hold the Orcale MySQL libraries anymore.

**Reply by demize** [reply]
> For x86_64:  
> Use the Alpine version which should have this plugin activated and enabled, the Debian version does not!

Unfortunately it looks like alpine isn't built with MySQL support at all (it's only compiled with postgres and sqlite). I was hoping that would be the solution, but alas.

The actual issue here isn't the lack of support for `caching_sha2_password` though, it's the regression where the server stops being able to connect to MySQL 8 servers entirely. I've set up my user with `mysql_native_password` and it still tries to use `caching_sha2_password` regardless; I've recreated the user entirely with the native_password type set from the start, changed the password type on the user after creation, regardless of that setting bitwarden_rs still tries to use `caching_sha2_password`. Per the original issue from this discussion, it worked fine in 1.17 and now doesn't work in 1.18 or 1.19, so something changed between those releases that caused a regression. 

I would be happy to use `mysql_native_password`, and I have tried it (multiple times), but using it doesn't work if the client won't try it. Unfortunately, this means right now I'm basically stuck with sqlite, which *works*, but I think we can agree is less than ideal.

**Comment by BlackDex** [comment]
Indeed I'm sorry, i have to correct that, the Alpine docker doesn't have mysql indeed.

**Comment by BlackDex** [comment]
I think this actually is something mysql server side when i read https://github.com/dani-garcia/bitwarden_rs/discussions/1348#discussioncomment-330624

Also, you could use MariaDB which does not have this issue instead of MySQL.

**Comment by demize** [comment]
It does *sound* like a server-side issue, but Seafile works fine using the same MariaDB client version (I imagine just major version; unfortunately, the nature of docker is such I have a hard time validating that) and Bitwarden worked fine prior to that 1.18 update (reportedly; I've been using an old version on Arch for a while, and only just tried to set it up using Docker, and I'd rather use sqlite for the time being than tie myself to a specific version).

It's a weird issue for sure... and if it's not something you changed yourself, it may be an upstream bug? That's probably the worst case, but I imagine it's the issue here because it's not sounding like anything you've changed yourself (and if it is, then my Seafile might stop working after an update... argh).

**Comment by BlackDex** [comment]
Well, Sealife uses Ubuntu 18.04 which packages there MariaDB libraries differently then Debian, and Ubuntu 18.04 (used by Sealife) seems to use version 10.1, while the debian version we are using is using 10.3. It could very well be a bug in Debian's building arguments.

I'm afraid we can't really do something about this at this moment. I am trying to get mysql to build on alpine with musl, but that hasn't been very successful as of yet since i haven't to much spare time to put into.

**Comment by timwis** [comment]
Hi all, I have been using bitwarden_rs just fine with the legacy authentication method (aka mysql_native_password), but that no longer works on the latest version of bitwarden_rs. I get the same error mentioned in the issue that this discussion was converted from:

> Plugin caching_sha2_password could not be loaded: /usr/lib/x86_64-linux-gnu/mariadb19/plugin/caching_sha2_password.so: cannot open shared object file: No such file or directory

I did some digging on this in https://github.com/dani-garcia/bitwarden_rs/issues/867 so I won't go into it again, but I would certainly file this as a regression. I will have to roll back to a previous version of bitwarden_rs in the meantime.

If I'm understanding correctly, bitwarden_rs does not work with a MySQL 8 server unless you change the server's default authentication method to the legacy authentication method (not just create a user that uses that authentication method, as @BlackDex suggests). As such, this does not work at all with Digital Ocean managed MySQL, or—I imagine—any other managed databases that use the latest version of MySQL.

**Comment by jjlin** [comment]
I don't think any of the regular contributors to bitwarden_rs actually use the MySQL backend, so if the MySQL users want a fix, someone will probably have to step up and figure out what's really going on. AFAICT, there have been no changes to the MySQL-related code or Rust dependencies since 1.17.0, so it's not at all clear to me that the regression is in bitwarden_rs. Perhaps it's on the Debian side, or perhaps it's on the server side.

**Comment by ttys3** [comment]
change your user 's auth method to the old one. or you need to read the doc on how to migration this.

sample mysql code:

```shell
 mysql -u root -p
Enter password: *********

mysql> ALTER USER 'root'@'localhost' IDENTIFIED WITH mysql_native_password BY 'newrootpassword';
Query OK, 0 rows affected (0.10 sec)

mysql> exit
```

For Windows 10:

Open the command prompt:

cd "C:\Program Files\MySQL\MySQL Server 8.0\bin"

C:\Program Files\MySQL\MySQL Server 8.0\bin> mysql -u root -p
Enter password: *********

mysql> ALTER USER 'root'@'localhost' IDENTIFIED WITH mysql_native_password BY 'newrootpassword';
Query OK, 0 rows affected (0.10 sec)

mysql> exit
Alternatively, you can change the my.ini configuration as the following:

```
[mysqld]

default_authentication_plugin=mysql_native_password
```

Restart the MySQL Server and open try restart the bidwarden container again.

by the way, you need know how to change the mysql container config.


refs:  https://stackoverflow.com/questions/49194719/authentication-plugin-caching-sha2-password-cannot-be-loaded

https://mysqlserverteam.com/mysql-8-0-4-new-default-authentication-plugin-caching_sha2_password/

by the way, there's nothing to do with docker version. it's about your mysql server config, and the container image version.

**Comment by everyx** [comment]
I just create a [mysql compat fork](https://github.com/everyx/bitwarden_rs/tree/feat/mysql8), use `libmysqlclient-dev` replace `libmariadb-dev` to make it compatible with mysql, and build a docker image [`everyx/bitwardenrs-server`](https://hub.docker.com/repository/docker/everyx/bitwardenrs-server/general) for this.

If anybody have the same problem, you can have a try for this.

**Reply by BlackDex** [reply]
I Actually think it's more the server side, which has that authentication enabled either by default with v8.0.x or enabled by something as default which causes this. So, just update/change the password to use the native way should be more then enough for this to be fixed.

One other item i can think about to maybe solve this is by adding the mariadb repo servers from mariadb them selfs to have newer versions available which could fix this, but not sure.

**Comment by Magneticdud** [comment]
Is it possible that this problem now also happens on mariadb? I am on mariadb 10.5.9 and I am getting this

edit: I accidentally connected to a mysql 8 database by inputting the wrong ip address. It works with mariadb

**Comment by BlackDex** [comment]
Also, Oracle MySQLv8 works fine for me using the default builds and images, just make sure you have the correct user password hashing method. See: https://github.com/dani-garcia/bitwarden_rs/wiki/FAQs#can-bitwarden_rs-connect-to-an-oracle-mysql-v8x-database
For more information.

**Comment by jjlin** [comment]
FYI, `caching_sha2_password.so` is shipping in Debian buster, as of June 19, 2021:

https://www.debian.org/News/2021/20210619.en.html#:~:text=caching_sha2_password.so

It's included in Vaultwarden 1.22.0 and newer Docker images.
